src/orion/algo/tpe.py

```diff
@@ -0,0 +1,396 @@
+# -*- coding: utf-8 -*-
+"""
+:mod:`orion.algo.tpe` -- Tree-structured Parzen Estimator Approach
+==================================================================
+
+.. module:: tpe
+   :platform: Unix
+   :synopsis: Tree-structured Parzen Estimator Approach
+
+"""
+import logging
+
+import numpy
+from scipy.stats import norm
+
+from orion.algo.base import BaseAlgorithm
+from orion.core.utils.points import flatten_dims, regroup_dims
+
+logger = logging.getLogger(__name__)
+
+
+def compute_max_ei_point(points, below_likelis, above_likelis):
+    """Compute ei among points based on their log likelihood and return the point with max ei.
+
+    :param points: list of point with real values.
+    :param below_likelis: list of log likelihood for each point in the good GMM.
+    :param above_likelis: list of log likelihood for each point in the bad GMM.
+    """
+    max_ei = -numpy.inf
+    point_index = 0
+    for i, (lik_b, lik_a) in enumerate(zip(below_likelis, above_likelis)):
+        ei = lik_b - lik_a
+        if ei > max_ei:
+            max_ei = ei
+            point_index = i
+    return points[point_index]
+
+
+# pylint:disable=assignment-from-no-return
+def adaptive_parzen_estimator(mus, low, high,
+                              prior_weight=1.0,
+                              equal_weight=False,
+                              flat_num=25):
+    """Return the sorted mus, the corresponding sigmas and weights with adaptive kernel estimator.
+
+    This adaptive parzen window estimator is based on the original papers and also refer the use of
+    prior mean in `this implementation
+    <https://github.com/hyperopt/hyperopt/blob/40bd617a0ca47e09368fb65919464ba0bfa85962/hyperopt/tpe.py#L400-L468>`_.
+
+    :param mus: list of real values for observed mus.
+    :param low: real value for lower bound of points.
+    :param high: real value for upper bound of points.
+    :param prior_weight: real value for the weight of the prior mean.
+    :param equal_weight: bool value indicating if all points with equal weights.
+    :param flat_num: int value indicating the number of latest points with equal weights,
+                     it is only valid if `equal_weight` is False.
+    """
+    def update_weights(total_num):
+        """Generate weights for all components"""
+        if total_num < flat_num or equal_weight:
+            return numpy.ones(total_num)
+
+        ramp_weights = numpy.linspace(1.0 / total_num, 1.0, num=total_num - flat_num)
+        flat_weights = numpy.ones(flat_num)
+        return numpy.concatenate([ramp_weights, flat_weights])
```

Comment:
From this paper: http://proceedings.mlr.press/v28/bergstra13.pdf

>  The first modification was to downweight trials as they age so that old results do not count for as much as more recent ones. We gave full weight to the most recent 25 trials and applied a linear ramp from 0 to 1.0 to older trials. This is a heuristic concession to TPE’s assumption of hyperparameter independence: as our search moves through space, we use temporal distance within the experiment as a surrogate for distance in the search space.

Seems what we are doing here is to ramp up from 1 / n up to 1.

Reply:
yeah, I notice this too, but seems hyperopt did this too even the paper claim the other way.

Reply:
Mismatch between paper and code in hyperband and now TPE again! :facepalm: I would be curious to compare both version, but now isn't a good time. Maybe when we start benchmarking the algos. Do you think it's worth adding a `paper_rampup` parameter so that we can test both versions later on? I think a pointer to the code would again be important to justify the discrepancy with the paper.

Reply:
For pointer, since we now put the weight ramp function inside the parzen estimator which will call the ramp function, and we already has a pointer for parzen to hyperopt. I guess we can omit this one. 

For the papar_rampup, I guess we can, let's add this later together with the discrete support, as I assume there will be some other changes.

Reply:
I believe it should be more explicit. We could re-use the same pointer but clearly state what we are referring to, that is, the use of an additional mid-space point and the ramp-up which is different than the paper, flat (and low) for trials before 25 last ones, and then linear up to 1.

Reply:
okay, let's make it more explicitly when we have the paper_rampup weight update function, when we should move the current implementation outside parzen estimator and user can choose.